quinn-proto/src/connection.rs

```diff
@@ -2073,6 +2081,28 @@ impl Connection {
         }
     }
 
+    fn push_crypto(&mut self, crypto: Crypto, number: u64) {
+        let old_idx = self
+            .cryptos
+            .iter()
+            .position(|(level, _, _)| level == &CryptoLevel::OneRtt)
+            .unwrap();
+        self.cryptos.remove(old_idx);
+        self.cryptos
+            .push_back((CryptoLevel::OneRtt, number, crypto));
```

Comment:
If this would cause three sets of 1-RTT keys to be known, we should drop the oldest as it will never be used (since `key_phase` can only address two sets of keys) and to prevent unbounded memory use wrt. time.

Reply:
Actually, I think this code has the opposite problem from what you suggest. It always discards the oldest `OneRtt` crypto when it pushes back an update, which I think means it will never retain more than one encryption level. (We should probably add a test for that case.)

Reply:
That's what I get for staying up late doing code review...

Reply:
Well, you triggered the finding of a bug after all -- not such a bad job.